### PR TITLE
gh-101152:  Implement PEP 699

### PR DESCRIPTION
**Pull request by ramvikrams**
gh-101152: Added the PEP 699 Implementation

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
1. Please add news.
2. The code looks similar to Kumar's. I'll add him as co-author when I merge.

**Comment by ramvikrams** [comment]
> 1. Please add news.
> 2. The code looks similar to Kumar's. I'll add him as co-author when I merge.

added the news and infact it is a replica of his solution

File: Misc/NEWS.d/next/Build/2023-01-21-10-31-35.gh-issue-101152.xvM8pL.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Added the PEP-699 Implementation
```
Comment (Fidget-Spinner) [inline Misc/NEWS.d/next/Build/2023-01-21-10-31-35.gh-issue-101152.xvM8pL.rst]: Try to be more descriptive. You should say what this means and what are the effects on external code. For example, how would 3rd party code be affected?

Also please add this to the whats new in 3.12 document.

Comment (ramvikrams) [reply Misc/NEWS.d/next/Build/2023-01-21-10-31-35.gh-issue-101152.xvM8pL.rst]: yes I'll add it

File: Doc/whatsnew/3.12.rst [hunk 2]
```diff
@@ -413,6 +413,10 @@ Deprecated
   is no current event loop set and it decides to create one.
   (Contributed by Serhiy Storchaka and Guido van Rossum in :gh:`100160`.)
 
+* Added the PEP-699 Implementation, :meth:`ma_version_tag` is deprecated for extension modules
```
Comment (Fidget-Spinner) [inline Doc/whatsnew/3.12.rst]: Is the `:meth:` markup correct? This isn't a method, it's a field/attribute. There's a guide here https://devguide.python.org/documentation/markup/.

File: Misc/NEWS.d/next/Build/2023-01-21-10-31-35.gh-issue-101152.xvM8pL.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Added the PEP-699 Implementation, :meth:`ma_version_tag` is deprecated for extension modules you can continue using it for the core code.
+The deprecation warning will pop up for extension modules.
```
Comment (Fidget-Spinner) [inline Misc/NEWS.d/next/Build/2023-01-21-10-31-35.gh-issue-101152.xvM8pL.rst]: Builtin CPython code should not use the field either. We shouldn't allow this by saying you "can continue using it".

I suggest the following,

```suggestion
In accordance with PEP 699, the ``ma_version_tag`` field in :c:type:`PyDictObject`
is deprecated for extension modules. Accessing this field will generate a compiler
warning at compile time. This field will be removed in Python 3.13.
```

File: Doc/whatsnew/3.12.rst [hunk 4]
```diff
@@ -413,6 +413,10 @@ Deprecated
...
+  you can continue using it for the core code. The deprecation warning will pop up for extension modules.
+  (Contributed by Ramvikrams and Kumar Aditya in :gh:`101193`.)
+
```
Comment (Fidget-Spinner) [inline Doc/whatsnew/3.12.rst]:
```suggestion
* In accordance with PEP 699, the ``ma_version_tag`` field in :c:type:`PyDictObject`
  is deprecated for extension modules. Accessing this field will generate a compiler
  warning at compile time. This field will be removed in Python 3.13.
  (Contributed by Ramvikrams and Kumar Aditya in :gh:`101193`. PEP by Ken Jin.)
```

**Review by Fidget-Spinner — COMMENTED** [review]
Just some feedback for your future contributions:

1. You can close conversations. Click on the "resolve conversation" button once you feel that you resolved the comment.
2. reStructuredText markup is a little tricky! So you can refer to the devguide for that https://devguide.python.org/documentation/markup/
3. I'll try to summarise what I look for in the Whats New deprecation. First you want to know exactly *what* is being affected (in this case the field of a certain struct being deprecated). Secondly, you want to know, without ambiguity, how we will notify users of this deprecation. In this case it's a compiler warning. Lastly, if there's a planned removal version for the deprecated thing, it should be mentioned too.

Thank you.

File: Doc/whatsnew/3.12.rst (hunk 2 above)
Comment (ramvikrams) [reply Doc/whatsnew/3.12.rst]: thanks, got the mistake

**Comment by ramvikrams** [comment]
> Just some feedback for your future contributions:

The feedback was very helpful, will surely keep the points in mind.

Thanks

File: Doc/whatsnew/3.12.rst [hunk 5]
```diff
@@ -413,6 +413,10 @@ Deprecated
   is no current event loop set and it decides to create one.
   (Contributed by Serhiy Storchaka and Guido van Rossum in :gh:`100160`.)
 
+* In accordance with PEP 699, the ``ma_version_tag`` field in :c:type:`PyDictObject`
+  is deprecated for extension modules. Accessing this field will generate a compiler
+  warning at compile time. This field will be removed in Python 3.13.
```
Comment (Fidget-Spinner) [inline Doc/whatsnew/3.12.rst]: Ah wait sorry, I forgot we're already in 3.12. So this should be changed to say removed in 3.14. Please update the news item as well.

Comment (ramvikrams) [reply Doc/whatsnew/3.12.rst]: yes i'll do that

File: Doc/whatsnew/3.12.rst [hunk 6]
```diff
@@ -413,6 +413,10 @@ Deprecated
   is no current event loop set and it decides to create one.
   (Contributed by Serhiy Storchaka and Guido van Rossum in :gh:`100160`.)
 
+* In accordance with PEP 699, the ``ma_version_tag`` field in :c:type:`PyDictObject`
```
Comment (Fidget-Spinner) [inline Doc/whatsnew/3.12.rst]: Woops, I also just realised we can markup the PEP to link to it. Sorry, could you adjust the news too?
```suggestion
* In accordance with :pep:`699`, the ``ma_version_tag`` field in :c:type:`PyDictObject`
```

**Review by kumaraditya303 — APPROVED** [review]
LGTM

**Comment by Fidget-Spinner** [comment]
Can you rebase with main and fix the conflicts in 3.12 whats new please? Seems like someone wrote an entry at the current spot.

**Comment by ramvikrams** [comment]
> Can you rebase with main and fix the conflicts in 3.12 whats new please? Seems like someone wrote an entry at the current spot.

yes sure

**Review by Fidget-Spinner — APPROVED** [review]

**Pull request merged**